### ISSUE TITLE
AssertionError in rule make_contigs_kmer_index

### ISSUE THREAD

**Issue by taylorreiter** [spacegraphcats/spacegraphcats#369]
```
rule make_contigs_kmer_index:
    input: rgnv_nbhd_k31/bcalm.unitigs.db
    output: rgnv_nbhd_k31_r5/contigs.mphf, rgnv_nbhd_k31_r5/contigs.indices
    jobid: 2


        /home/tereiter/miniconda3/envs/sgc_dom_abund/bin/python -m spacegraphcats.cdbg.index_cdbg_by_kmer             -k 31 rg
nv_nbhd_k31_r5 --contigs-db rgnv_nbhd_k31/bcalm.unitigs.db

/home/tereiter/miniconda3/envs/sgc_dom_abund/lib/python3.8/runpy.py:127: RuntimeWarning: 'spacegraphcats.cdbg.index_cdbg_by_km
er' found in sys.modules after import of package 'spacegraphcats.cdbg', but prior to execution of 'spacegraphcats.cdbg.index_c
dbg_by_kmer'; this may result in unpredictable behaviour
  warn(RuntimeWarning(msg))
reading cDBG nodes from sqlite DB rgnv_nbhd_k31/bcalm.unitigs.db
loaded 46667471 contigs.

building MPHF for 519194226 k-mers in 46667471 nodes.
second pass.
reading cDBG nodes from sqlite DB rgnv_nbhd_k31/bcalm.unitigs.db
loaded 46667471 contigs in pass2.

Traceback (most recent call last):
  File "/home/tereiter/miniconda3/envs/sgc_dom_abund/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/home/tereiter/miniconda3/envs/sgc_dom_abund/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/tereiter/miniconda3/envs/sgc_dom_abund/lib/python3.8/site-packages/spacegraphcats/cdbg/index_cdbg_by_kmer.py", l
ine 216, in <module>
    sys.exit(main(sys.argv[1:]))
  File "/home/tereiter/miniconda3/envs/sgc_dom_abund/lib/python3.8/site-packages/spacegraphcats/cdbg/index_cdbg_by_kmer.py", l
ine 205, in main
    table, sizes = build_mphf(a.ksize, create_records_iter)
  File "/home/tereiter/miniconda3/envs/sgc_dom_abund/lib/python3.8/site-packages/spacegraphcats/cdbg/index_cdbg_by_kmer.py", l
ine 181, in build_mphf
    assert n == max(table.mphf_to_value), (n, max(table.mphf_to_value))
AssertionError: (46667470, 4294967295)
```

When I've had errors like this in the past, some earlier sgc steps have silently failed due to insufficient RAM, but snakemake has continued running. I can't assess whether that is the problem here because i ran this in an srun session and can't scroll back that far. sigh. 

I installed sgc with:
```
conda env create -f sgc_env.yml -n sgc_dom_abund
conda activate sgc_dominator_abund
pip install git+https://github.com/spacegraphcats/spacegraphcats@dominator_abund
```

where `sgc_env.yml` was

```
name: sgc_dom_abund
channels:
    - conda-forge
    - bioconda
    - defaults
dependencies:
    - bcalm=2.2.3
    - snakemake-minimal=5.20.1
    - screed=1.0.1
    - khmer=3.0.0a3
    - numpy
    - pytest=5.3.2
    - sortedcontainers=2.1.0
    - click=7.1.2
    - pyflakes
    - bbhash>=0.5.4
    - pip
    - pip:
      - sourmash==4.0.0a3
```

I'm also attaching my conf file:
[rgnv_nbhd_catlas_r5_conf.yml.txt](https://github.com/spacegraphcats/spacegraphcats/files/5667322/rgnv_nbhd_catlas_r5_conf.yml.txt)

**Comment by taylorreiter** [comment]
Maybe related to #249 but hard for me to assess.

**Comment by ctb** [comment]
This is not related to earlier failures. I vaguely recall running into this and spending a lot of time debugging it - will dig in again!

**Comment by taylorreiter** [comment]
The branch I installed sgc from is 7 commits behind `latest`...could this be fixed by merging latest in?

**Comment by ctb** [comment]
On Wed, Dec 09, 2020 at 08:52:54AM -0800, Taylor Reiter wrote:
> The branch I installed sgc from is 7 commits behind `latest`...could this be fixed by merging latest in?

yes.

**Comment by ctb** [comment]
(I mean, it might not be, but I do recall figuring this out at some point.)

What version of spacegraphcats do you have installed? `pip show spacegraphcats` should give you that.

**Comment by taylorreiter** [comment]
```
$ pip show spacegraphcats
Name: spacegraphcats
Version: 2.0b8.dev31+g976e192
Summary: tools for biological assembly graph neighborhood analysis
Home-page: https://github.com/spacegraphcats/spacegraphcats
Author: C. Titus Brown, Dominik Moritz, Michael P. O'Brien, Felix Reidl, Taylor Reiter, and Blair D. Sullivan
Author-email: titus@idyll.org,blair.d.sullivan@gmail.com
License: BSD 3-clause
Location: /home/tereiter/miniconda3/envs/sgc_dom_abund/lib/python3.8/site-packages
Requires: bbhash, pytest, screed, sortedcontainers, khmer, sourmash, numpy, snakemake
```

**Comment by ctb** [comment]
hmm. I don't see anything in the commit logs that would have fixed this issue, but if you could try rerunning from scratch that would be ideal and would at least rule out that as a possibility. thanks!

**Comment by taylorreiter** [comment]
okey dokey will do!

**Comment by taylorreiter** [comment]
Same exact error message, this time with `This is spacegraphcats 2.0b11.`

```
rule make_contigs_kmer_index:
    input: rgnv_nbhd_k31/bcalm.unitigs.db
    output: rgnv_nbhd_k31_r5/contigs.mphf, rgnv_nbhd_k31_r5/contigs.indices
    jobid: 2


This is spacegraphcats 2.0b11.

Job counts:
        count   jobs
        1       make_contigs_kmer_index
        1
/home/tereiter/miniconda3/envs/sgc_dom_abund/lib/python3.8/runpy.py:127: RuntimeWarning: 'spacegraphcats.cdbg.index_cdbg_by_kmer' foun$
 in sys.modules after import of package 'spacegraphcats.cdbg', but prior to execution of 'spacegraphcats.cdbg.index_cdbg_by_kmer'; thi$
 may result in unpredictable behaviour
  warn(RuntimeWarning(msg))
reading cDBG nodes from sqlite DB rgnv_nbhd_k31/bcalm.unitigs.db
loaded 46667471 contigs.

building MPHF for 519194226 k-mers in 46667471 nodes.
second pass.
reading cDBG nodes from sqlite DB rgnv_nbhd_k31/bcalm.unitigs.db
loaded 46667471 contigs in pass2.

Traceback (most recent call last):
  File "/home/tereiter/miniconda3/envs/sgc_dom_abund/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/home/tereiter/miniconda3/envs/sgc_dom_abund/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/tereiter/miniconda3/envs/sgc_dom_abund/lib/python3.8/site-packages/spacegraphcats/cdbg/index_cdbg_by_kmer.py", line 216,
in <module>
    sys.exit(main(sys.argv[1:]))
  File "/home/tereiter/miniconda3/envs/sgc_dom_abund/lib/python3.8/site-packages/spacegraphcats/cdbg/index_cdbg_by_kmer.py", line 205,
in main
    table, sizes = build_mphf(a.ksize, create_records_iter)
  File "/home/tereiter/miniconda3/envs/sgc_dom_abund/lib/python3.8/site-packages/spacegraphcats/cdbg/index_cdbg_by_kmer.py", line 181,
in build_mphf
    assert n == max(table.mphf_to_value), (n, max(table.mphf_to_value))
AssertionError: (46667470, 4294967295)
```

**Comment by ctb** [comment]
ok, dug into this a bit this morning. This will be a little bit train-of-thought-y.

First, the number `4294967295` is important. This is `2**32 - 1`, or the max unsigned integer UINT32_MAX. In the context of the pybbhash table that is being built here, it indicates that there is a k-mer whose associated value in the table is not set (and so is set to the default value, UINT32_MAX - 1).

Backing up a bit, this bit of code is doing the following:

* walk over all cDBG unitigs in the `rgnv_nbhd_k31/bcalm.unitigs.db` and turn them into hash values
* build an Minimal Perfect Hash Function encoding all observed hash values into integers `0..max(hashval)`
* build a table of size `max(hashval)`
* go back over all cDBG unitigs and connect each hash value to unitig_id in the table by setting `table[mphf(hashval)] = unitig_id`

This error says that somewhere, somehow, there is a hash value in the table that does not have a unitig_id set.

Needless to say, this should never happen :).

To debug this, I'd like to have access to the data - presumably this is on farm, could you drop me the location? No hurry.

Also, per our conversation yesterday, this is independent of the radius; this code only operates on the cDBG/bcalm output.

(I'll update the assert with a bit more documentation, too.)

**Comment by taylorreiter** [comment]
files on cluster: 
conf file lives here: `/home/tereiter/github/2020-ibd/sandbox/test_sgc_dominator_abund/sgc_nbhd_rgnv_r5`
I made all of the other paths relative to that directory, but the path to those files is (all one line):
`/home/tereiter/github/2020-ibd/outputs/sgc_genome_queries/*_k31_r1_search_oh0/GCF_900036035.1_RGNV35913_genomic.fna.gz.cdbg_ids.reads.fa.gz`

**Comment by ctb** [comment]
running now, thx!

**Comment by ctb** [comment]
"Fixed" in #375!

Note that the r5 catlas is extremely expensive to build, I'm up to 400 GB and still can't do it ;).

**Comment by taylorreiter** [comment]
Yay!! And yes, accurate...I used 850gb of ram prior to the sgc rewrite.
